domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/CpuPerformanceSnapshotter.kt

```diff
@@ -0,0 +1,209 @@
+package org.oppia.android.domain.oppialogger.analytics
+
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.channels.actor
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.oppia.android.app.model.ScreenName
+import org.oppia.android.util.logging.ConsoleLogger
+import org.oppia.android.util.logging.ExceptionLogger
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor.AppIconification
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor.Snapshot
+import org.oppia.android.util.threading.BackgroundDispatcher
+import javax.inject.Inject
+
+/**
+ * Snapshotter that gracefully and sequentially logs cpu usage across foreground and background
+ * moves of the application.
+ */
+class CpuPerformanceSnapshotter private constructor(
+  private val backgroundCoroutineDispatcher: CoroutineDispatcher,
+  private val performanceMetricsLogger: PerformanceMetricsLogger,
+  private val initialIconification: AppIconification,
+  private val consoleLogger: ConsoleLogger,
+  private val exceptionLogger: ExceptionLogger,
+  private val foregroundCpuLoggingTimePeriodMillis: Long,
+  private val backgroundCpuLoggingTimePeriodMillis: Long,
+  private val performanceMetricsAssessor: PerformanceMetricsAssessor
+) {
+
+  private val commandQueue by lazy { createCommandQueueActor() }
+
+  /** Updates the current [AppIconification] in accordance with the app's state changes. */
+  fun updateAppIconification(newIconification: AppIconification) {
+    sendSwitchIconificationCommand(newIconification)
+  }
+
+  private fun createCommandQueueActor(): SendChannel<CommandMessage> {
+    var currentIconification = initialIconification
+    var previousSnapshot =
+      performanceMetricsAssessor.computeSnapshotAtCurrentTime(currentIconification)
+    var switchIconificationCount = 0
+    val coroutineScope = CoroutineScope(backgroundCoroutineDispatcher)
+    return coroutineScope.actor<CommandMessage>(capacity = Channel.UNLIMITED) {
+      for (message in channel) {
+        when (message) {
+          is CommandMessage.SwitchIconification -> {
+            ++switchIconificationCount
+            sendLogSnapshotDiffCommand(
```

Comment:
It's probably worth documenting why this bit is needed, e.g. that we want to capture whatever's been running up to the background/foreground transition and not lose that (even though the normal full period of time may have not yet completed).

Reply:
nice idea, done.